### PR TITLE
8439 Coop IA Document validations

### PR DESCRIPTION
**Pull request by cameron-freshworks**
*Issue #:* /bcgov/entity#8439

*Description of changes:*
* Implement document property and file size checks in IA Validators.
* Updated unit tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py [hunk 1]
```diff
@@ -226,6 +234,59 @@ def validate_incorporation_effective_date(incorporation_json) -> Error:
...
+
+    if not memorandum_file_name:
+        msg.append({'error': babel('A valid memorandum file name is required.')})
+
+    if msg:
+        return msg
+
+    # Validate the files from Minio
+    try:
+        MinioService.get_file(rules_file_key)
```
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/services/filings/validations/incorporation_application.py:270]: Do we need this get?

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py [hunk 2]
```diff
@@ -226,6 +234,59 @@ def validate_incorporation_effective_date(incorporation_json) -> Error:
...
+        file_info = MinioService.get_file_stat_object(rules_file_key)
+        if file_info.size > 10000:
+            msg.append({'error': babel('Rules file exceeds maximum size.')})
+    except Exception:
+        msg.append({'error': babel('Invalid rules file.')})
+
+    try:
+        memorandum_file = MinioService.get_file(memorandum_file_key)
```
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/services/filings/validations/incorporation_application.py:278]: Do we need this get?

**Review by vysakh-menon-aot — CHANGES_REQUESTED** [review]

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 1 above)
Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:270]: We need to get the file to validate that the document page size is letter size.  You are going to add that code right Cam?

Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:270]: Yea, I’ve genuinely struggled with this though. I haven’t figured out how to do this server side as of yet. Could def use some help here!

File: legal-api/src/legal_api/services/minio.py [hunk 3]
```diff
@@ -70,6 +70,12 @@ def get_file(key: str):
         minio_client: Minio = MinioService._get_client()
         return minio_client.get_object(current_app.config['MINIO_BUCKET_LEAR'], key)
 
+    @staticmethod
+    def get_file_stat_object(key: str):
+        """Fetch file stat object from Minio."""
+        minio_client: Minio = MinioService._get_client()
+        return minio_client.stat_object(current_app.config['MINIO_BUCKET_LEAR'], key)
```
[the remark this answers is not in the document]
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/minio.py:77]: Oh right on.  I’ll check it out

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 1 above)
Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:270]: yeah we can talk about this

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 2 above)
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:278]: We will once i figure out how to read/verify this 👍

**Comment by cameron-freshworks** [comment]
Setting to draft while i update the document reading and fix linting/tests.

**Comment by cameron-freshworks** [comment]
> Setting to draft while i update the document reading and fix linting/tests.

Still a few more of these to fix, on it 👍

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py [hunk 4]
```diff
@@ -47,9 +51,15 @@ def validate(incorporation_json: Dict):
...
-        msg.extend(err)
+    if legal_type == Business.LegalTypes.BCOMP.value:
+        err = validate_share_structure(incorporation_json, coreFiling.FilingTypes.INCORPORATIONAPPLICATION.value)
+        if err:
+            msg.extend(err)
+
+    if legal_type == Business.LegalTypes.COOP.value:
+        err = validate_cooperative_documents(incorporation_json)
+        if err:
+            msg.extend(err)
```
Comment (argush3) [inline legal-api/src/legal_api/services/filings/validations/incorporation_application.py:63]: maybe we should `elsif` here in preparation for other legal types.

Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:63]: For sure, will do!

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py [hunk 5]
```diff
@@ -227,6 +237,48 @@ def validate_incorporation_effective_date(incorporation_json) -> Error:
...
+    if msg:
+        return msg
+
+    rules_err = validate_pdf(rules_file_key)
+    if rules_err:
+        return rules_err
+
+    memorandum_err = validate_pdf(memorandum_file_key)
+    if memorandum_err:
+        return memorandum_err
```
Comment (argush3) [inline legal-api/src/legal_api/services/filings/validations/incorporation_application.py:277]: potentially update so you can return all validation errors in one go?  maybe keep have a `rules_files_exists` and `memorandum_file_exists` flag to decide whether to call the `validate_pdf` functions.

File: legal-api/tests/unit/services/filings/validations/test_incorporation_application.py [hunk 6]
```diff
@@ -593,3 +609,119 @@ def test_validate_incorporation_effective_date(session, test_name, effective_dat
...
+        if scenario == 'invalidRulesSize':
+            f['filing']['incorporationApplication']['cooperative']['rulesFileKey'] = _upload_file(legal)
+            f['filing']['incorporationApplication']['cooperative']['memorandumFileKey'] = _upload_file(letter)
+        if scenario == 'invalidMemorandumSize':
+            f['filing']['incorporationApplication']['cooperative']['rulesFileKey'] = _upload_file(letter)
+            f['filing']['incorporationApplication']['cooperative']['memorandumFileKey'] = _upload_file(legal)
+    else:
+        # Assign key and value to test empty variables for failures
+        keyValue = ''
+        f['filing']['incorporationApplication']['cooperative'][key] = keyValue
```
Comment (argush3) [inline legal-api/tests/unit/services/filings/validations/test_incorporation_application.py:684]: nice, very click

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 5 above)
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:277]: Absolutely, will try!

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py [hunk 7]
```diff
@@ -292,3 +344,28 @@ def validate_correction_name_request(filing: Dict, corrected_filing: Dict) -> Op
         return msg
 
     return None
+
+
+def validate_pdf(file_key: str):
```
Comment (argush3) [inline legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: just remembered we need to check that the file isn't encrypted

**Comment by argush3** [comment]
looks good!  the encryption thing is the main thing we need to get in here.  will approve once that's in.

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 7 above)
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: ah geez, not sure how to do that.

Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: probly super easy.  let me try and find the code.

Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: just use `pdf_reader.isEncrypted` to check

Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: Thats way too simple.. can't be real :)

Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: yeah i know hey 😀

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 5 above)
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:277]: So i'm unsure about this one. I do like the accurate error messages without them adding up.

If we fail to have a key value we return with the msg and don't validate the form in the scenario of a bad key. 
From the refactoring i've tried we end up with more conditions if we want to accumulate everything before returning.  
Are you ok with how it currently is?

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 7 above)
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: Added 👍

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 5 above)
Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:277]: yeah that's fine.

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 7 above)
Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: do you think "File must be unencrypted" is more ideal for the validation message?

Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: Yes, thats a little for sure! Updated 👍

Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:349]: looks good!

**Comment by andrepestana-aot** [comment]
Looks good to me! Just check the code smells pointed by sonar.

**Review by argush3 — APPROVED** [review]

**Review by vysakh-menon-aot — APPROVED** [review]

**Pull request merged**